### PR TITLE
bpo-40958: Avoid buffer overflow in the parser when indexing the current line

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue40958

### CODE REVIEW

**Comment by pablogsal** [comment]
With this PR:

```
venv ❯ LSAN_OPTIONS="suppressions=asan-suppression.txt,print_suppressions=0" ./python -m test test_eof
0:00:00 load avg: 1.13 Run tests sequentially
0:00:00 load avg: 1.13 [1/1] test_eof

== Tests result: SUCCESS ==

1 test OK.

Total duration: 493 ms
Tests result: SUCCESS
^[[A
~/github/python/master bpo-40958*
venv ❯ LSAN_OPTIONS="suppressions=asan-suppression.txt,print_suppressions=0" ./python -c '"Ṕýţĥòñ" +'
  File "<string>", line 1
    "Ṕýţĥòñ" +
              ^
SyntaxError: invalid syntax

~/github/python/master bpo-40958*
venv ❯ LSAN_OPTIONS="suppressions=asan-suppression.txt,print_suppressions=0" ./python -c 'yield from'
  File "<string>", line 1
    yield from
              ^
SyntaxError: invalid syntax
```

**Review by tiran — CHANGES_REQUESTED** [review]
A test is failing:
```
 ======================================================================
FAIL: testSyntaxErrorOffset (test.test_exceptions.ExceptionTests) (source=b'Python = "\xcf\xb3\xf2\xee\xed" +', lineno=1, offset=18)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\cpython\cpython\lib\test\test_exceptions.py", line 188, in check
    self.assertEqual(cm.exception.offset, offset)
AssertionError: 13 != 18
```

File: Parser/pegen.c [hunk 1]
```diff
@@ -141,7 +141,8 @@ _create_dummy_identifier(Parser *p)
 static inline Py_ssize_t
 byte_offset_to_character_offset(PyObject *line, int col_offset)
 {
-    const char *str = PyUnicode_AsUTF8(line);
+    Py_ssize_t linesize;
+    const char *str = PyUnicode_AsUTF8AndSize(line, &linesize);
```
Comment (lysnikolaou) [inline Parser/pegen.c:145]: Why does this make a difference, since `linesize` is not used anywhere thereafter?

Comment (pablogsal) [reply Parser/pegen.c:145]: This was badly committed, give me a sec because I am still missing something

Comment (lysnikolaou) [reply Parser/pegen.c:145]: Oh, sorry!

**Comment by pablogsal** [comment]
Closing this as this is missing something important and unfortunately, I don't have time to investigate today :(

**Comment by pablogsal** [comment]
@lysnikolaou Feel free to investigate if you have some time

**Comment by lysnikolaou** [comment]
> @lysnikolaou Feel free to investigate if you have some time

I'll do as soon as I get home. Sorry for not being able to help out now!

**Pull request closed without merging**